consult.el

```diff
@@ -723,35 +750,74 @@ The line beginning/ending BEG/END is bound in BODY."
-If DIR is a true value, the user is asked.
-Then the `consult-project-function' is tried.
-Otherwise the `default-directory' is returned."
-  (let* ((dir
+If DIR is null or equal to `default-directory', local
+`search-path-list' is set to nil.  Otherwise, when DIR
+is a string, it is split into space-separated path
+patterns, pathname wildcard expansion is performed on
+each element and a flattened list of the results is
+saved in `search-path-list'.
+
+The last element of the return value is either the
+expanded value of DIR if it is a single directory, the
+current project root directory if any from a call to
+`consult-project-function' or else `default-directory'."
+  (let* ((abbrev-dir ".")
+	 (possible-prefix-arg dir)
+	 (dir
           (cond
-           ((stringp dir) dir)
+           ((or (stringp dir) (and (listp dir) (stringp (car dir))))
+	    dir)
            (dir
             ;; Preserve this-command across `read-directory-name' call,
             ;; such that `consult-customize' continues to work.
             (let ((this-command this-command))
-              (read-directory-name "Directory: " nil nil t)))
+	      (pcase dir
+		;; Single C-u prefix arg, prompt for a single directory
+		('(4) (read-directory-name "Directory: " nil nil t))
+		;; Any other prefix arg, prompt for any number of directories
+		;; (can be zero) followed by any number of path patterns with
+		;; wildcards
+		(_    (read-string "Search Locations (default = .): "
+				   nil nil abbrev-dir)))))
            (t (or (consult--project-root) default-directory))))
-         (edir (file-name-as-directory (expand-file-name dir)))
+         (edir (file-name-as-directory
+		(expand-file-name (if (and (stringp dir)
+					   (file-directory-p dir))
+				      dir
+				    abbrev-dir))))
          ;; Bind default-directory in order to find the project
-         (pdir (let ((default-directory edir)) (consult--project-root))))
-    (cons
-     (format "%s (%s): " prompt
-             (if (equal edir pdir)
-                 (concat "Project " (consult--project-name pdir))
-               (consult--abbreviate-directory dir)))
+         (pdir (let ((default-directory edir)) (consult--project-root)))
+	 (search-path-list
+	  (if (equal dir default-directory)
+	      (list abbrev-dir)
+            (when (stringp dir)
+	      ;; Expand any space separated list of wildcard path patterns
+	      (flatten-list
+	       (mapcar (lambda (pat) (or (file-expand-wildcards pat) pat))
+		       (split-string dir)))))))
+    (list
+     (format "%s %s: " prompt
+	     (cond ((and possible-prefix-arg (not (stringp possible-prefix-arg))
+			 (not (eq possible-prefix-arg '(4))))
+		    (if (listp dir)
+			dir
+		      (format "(%s)" dir)))
+		   ((equal edir pdir)
+                    (format "(Project %s)" (consult--project-name pdir)))
+		   (t (format "(%s)"
+			      (consult--abbreviate-directory
+			       (if (and (stringp dir) (file-directory-p dir))
+				   dir
+				 abbrev-dir))))))
```

Comment:
I must admit I don't understand what you are doing here. What do you want to achieve? How should the prompt look for directory search, project search and wildcard search? I propose the following:

- `Grep (Project consult): `
- `Grep (some/abbreviated/dir): `
- `Grep (first/abbreviated/unexpanded/wildcard, …): `

Reply:
I was looking for and getting pretty much what you are showing except that the I did not deal with abbreviating all the paths.  I'll look at it.

Reply:
Did my best but the conditional logic here requires some complexity.

Reply:
I fiddled a bit around with this and it is possible to reduce the conditional logic significantly. But this does not matter much. We should first discuss the wildcard expansion issue, I mentioned in https://github.com/minad/consult/pull/749#issuecomment-1445043576.